File: clients/jvm/src/main/kotlin/com/kgd/client/Protocol.kt
package com.kgd.client

import org.msgpack.core.MessagePack
import org.msgpack.core.MessagePacker
import org.msgpack.core.MessageUnpacker
import org.msgpack.value.Value
import org.msgpack.value.ValueFactory
import java.io.ByteArrayOutputStream
import java.io.InputStream

/**
 * msgpack-RPC message types.
 */
internal object MsgType {
    const val REQUEST = 0
    const val RESPONSE = 1
    const val NOTIFICATION = 2
}

/**
 * RPC method names.
 */
internal object Method {
    const val HELLO = "hello"
    const val UPLOAD = "upload"
    const val PLACE = "place"
    const val UNPLACE = "unplace"
    const val UNPLACE_ALL = "unplace_all"
    const val FREE = "free"
    const val REGISTER_WIN = "register_win"
    const val UPDATE_SCROLL = "update_scroll"
    const val UNREGISTER_WIN = "unregister_win"
    const val LIST = "list"
    const val STATUS = "status"
    const val STOP = "stop"
}

/**
 * Server-to-client notification names.
 */
internal object Notify {
    const val EVICTED = "evicted"
    const val TOPOLOGY_CHANGED = "topology_changed"
    const val VISIBILITY_CHANGED = "visibility_changed"
    const val THEME_CHANGED = "theme_changed"
}

/**
 * Represents a decoded msgpack-RPC message.
 */
internal sealed class RpcMessage {
    data class Response(val msgId: Int, val error: Value?, val result: Value?) : RpcMessage()
    data class Notification(val method: String, val params: List<Value>) : RpcMessage()
}

/**
 * Encodes msgpack-RPC request and notification messages.
 */
internal object Encoder {

    /**
     * Encode a request: `[0, msgid, method, [params]]`
     */
    fun encodeRequest(msgId: Int, method: String, params: Map<String, Any>?): ByteArray {
        val out = ByteArrayOutputStream()
        val packer = MessagePack.newDefaultPacker(out)
        packer.packArrayHeader(4)
        packer.packInt(MsgType.REQUEST)
        packer.packInt(msgId)
        packer.packString(method)
        if (params != null) {
            packer.packArrayHeader(1)
            packMap(packer, params)
        } else {
            packer.packArrayHeader(0)
        }
        packer.flush()
        return out.toByteArray()
    }

    /**
     * Encode a notification: `[2, method, [params]]`
     */
    fun encodeNotification(method: String, params: Map<String, Any>?): ByteArray {
        val out = ByteArrayOutputStream()
        val packer = MessagePack.newDefaultPacker(out)
        packer.packArrayHeader(3)
        packer.packInt(MsgType.NOTIFICATION)
        packer.packString(method)
        if (params != null) {
            packer.packArrayHeader(1)
            packMap(packer, params)
        } else {
            packer.packArrayHeader(0)
        }
        packer.flush()
        return out.toByteArray()
    }

    @Suppress("UNCHECKED_CAST")
    private fun packMap(packer: MessagePacker, map: Map<String, Any>) {
        packer.packMapHeader(map.size)
        for ((key, value) in map) {
            packer.packString(key)
            packValue(packer, value)
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun packValue(packer: MessagePacker, value: Any) {
        when (value) {
            is Int -> packer.packInt(value)
            is Long -> packer.packLong(value)
            is Boolean -> packer.packBoolean(value)
            is String -> packer.packString(value)
            is ByteArray -> {
                packer.packBinaryHeader(value.size)
                packer.writePayload(value)
            }
            is Map<*, *> -> packMap(packer, value as Map<String, Any>)
            is List<*> -> {
                packer.packArrayHeader(value.size)
                for (item in value) {
                    if (item != null) packValue(packer, item)
                    else packer.packNil()
                }
            }
            else -> packer.packNil()
        }
    }
}

/**
 * Streaming decoder for msgpack-RPC messages from an [InputStream].
 *
 * Call [readMessage] in a loop to receive parsed [RpcMessage] objects.
 */
internal class Decoder(input: InputStream) {
    private val unpacker: MessageUnpacker = MessagePack.newDefaultUnpacker(input)

    /**
     * Read and decode the next msgpack-RPC message.
     * Returns null on EOF or if the message cannot be parsed.
     * Throws [java.io.IOException] on read errors.
     */
    fun readMessage(): RpcMessage? {
        if (!unpacker.hasNext()) return null

        val msg = unpacker.unpackValue()
        if (!msg.isArrayValue) return null
        val arr = msg.asArrayValue().list()
        if (arr.size < 3) return null

        return when (arr[0].asIntegerValue().toInt()) {
            MsgType.RESPONSE -> {
                if (arr.size < 4) return null
                val msgId = arr[1].asIntegerValue().toInt()
                val error = if (arr[2].isNilValue) null else arr[2]
                val result = if (arr[3].isNilValue) null else arr[3]
                RpcMessage.Response(msgId, error, result)
            }
            MsgType.NOTIFICATION -> {
                val method = arr[1].asStringValue().toString()
                val paramsVal = arr[2]
                val params = if (paramsVal.isArrayValue) paramsVal.asArrayValue().list() else emptyList()
                RpcMessage.Notification(method, params)
            }
            else -> null
        }
    }

    fun close() {
        unpacker.close()
    }
}

/**
 * Helpers to convert msgpack [Value] objects into Kotlin types.
 */
internal object ValueHelper {

    fun asMap(value: Value?): Map<String, Value> {
        if (value == null || !value.isMapValue) return emptyMap()
        return value.asMapValue().entrySet().associate { (k, v) ->
            k.asStringValue().toString() to v
        }
    }

    fun asString(value: Value?, default: String = ""): String {
        if (value == null || value.isNilValue) return default
        return value.asStringValue().toString()
    }

    fun asInt(value: Value?, default: Int = 0): Int {
        if (value == null || value.isNilValue) return default
        return value.asIntegerValue().toInt()
    }

    fun asLong(value: Value?, default: Long = 0L): Long {
        if (value == null || value.isNilValue) return default
        return value.asIntegerValue().toLong()
    }

    fun asBool(value: Value?, default: Boolean = false): Boolean {
        if (value == null || value.isNilValue) return default
        return value.asBooleanValue().boolean
    }

    fun asList(value: Value?): List<Value> {
        if (value == null || !value.isArrayValue) return emptyList()
        return value.asArrayValue().list()
    }

    fun colorFromMap(map: Map<String, Value>): Color {
        return Color(
            r = asInt(map["r"]),
            g = asInt(map["g"]),
            b = asInt(map["b"]),
        )
    }
}

File: clients/swift/Sources/KgdClient/Protocol.swift
/// Msgpack-RPC protocol encoding and decoding for kgd.
///
/// Wire format:
/// - Request:      [0, msgid, method, [params]]
/// - Response:     [1, msgid, error, result]
/// - Notification: [2, method, [params]]

import Foundation
import MessagePack

// MARK: - Message types

/// Msgpack-RPC message type tags.
enum MsgType: Int {
    case request = 0
    case response = 1
    case notification = 2
}

// MARK: - RPC method names

enum Method {
    static let hello = "hello"
    static let upload = "upload"
    static let place = "place"
    static let unplace = "unplace"
    static let unplaceAll = "unplace_all"
    static let free = "free"
    static let registerWin = "register_win"
    static let updateScroll = "update_scroll"
    static let unregisterWin = "unregister_win"
    static let list = "list"
    static let status = "status"
    static let stop = "stop"
}

// MARK: - Server notification names

enum Notification {
    static let evicted = "evicted"
    static let topologyChanged = "topology_changed"
    static let visibilityChanged = "visibility_changed"
    static let themeChanged = "theme_changed"
}

// MARK: - Parsed messages

/// A parsed response from the server.
struct RPCResponse {
    let msgID: UInt32
    let error: MessagePackValue
    let result: MessagePackValue
}

/// A parsed notification from the server.
struct RPCNotification {
    let method: String
    let params: MessagePackValue
}

/// A parsed incoming message.
enum IncomingMessage {
    case response(RPCResponse)
    case notification(RPCNotification)
}

// MARK: - Encoding

/// Encode a request message: [0, msgid, method, [params]]
func encodeRequest(msgID: UInt32, method: String, params: MessagePackValue?) -> Data {
    let paramsArray: MessagePackValue
    if let p = params {
        paramsArray = .array([p])
    } else {
        paramsArray = .array([])
    }
    let msg: MessagePackValue = .array([
        .uint(UInt64(MsgType.request.rawValue)),
        .uint(UInt64(msgID)),
        .string(method),
        paramsArray,
    ])
    return pack(msg)
}

/// Encode a notification message: [2, method, [params]]
func encodeNotification(method: String, params: MessagePackValue?) -> Data {
    let paramsArray: MessagePackValue
    if let p = params {
        paramsArray = .array([p])
    } else {
        paramsArray = .array([])
    }
    let msg: MessagePackValue = .array([
        .uint(UInt64(MsgType.notification.rawValue)),
        .string(method),
        paramsArray,
    ])
    return pack(msg)
}

// MARK: - Decoding

/// Errors during message parsing.
enum ProtocolError: Error, CustomStringConvertible {
    case malformedMessage(String)
    case unknownMessageType(Int)

    var description: String {
        switch self {
        case .malformedMessage(let detail):
            return "malformed message: \(detail)"
        case .unknownMessageType(let t):
            return "unknown message type: \(t)"
        }
    }
}

/// Parse a single msgpack value into an IncomingMessage.
func parseMessage(_ value: MessagePackValue) throws -> IncomingMessage {
    guard let arr = value.arrayValue, arr.count >= 3 else {
        throw ProtocolError.malformedMessage("expected array with at least 3 elements")
    }

    guard let typeVal = arr[0].int64Value ?? arr[0].uint64Value.map({ Int64($0) }) else {
        throw ProtocolError.malformedMessage("message type is not an integer")
    }

    let msgType = Int(typeVal)

    switch msgType {
    case MsgType.response.rawValue:
        guard arr.count >= 4 else {
            throw ProtocolError.malformedMessage("response needs 4 elements")
        }
        guard let msgID = arr[1].uint64Value ?? arr[1].int64Value.map({ UInt64($0) }) else {
            throw ProtocolError.malformedMessage("response msgid is not an integer")
        }
        return .response(RPCResponse(
            msgID: UInt32(msgID),
            error: arr[2],
            result: arr[3]
        ))

    case MsgType.notification.rawValue:
        guard let method = arr[1].stringValue else {
            throw ProtocolError.malformedMessage("notification method is not a string")
        }
        let params: MessagePackValue
        if arr.count >= 3, let paramsArr = arr[2].arrayValue, !paramsArr.isEmpty {
            params = paramsArr[0]
        } else {
            params = .nil
        }
        return .notification(RPCNotification(method: method, params: params))

    default:
        throw ProtocolError.unknownMessageType(msgType)
    }
}

// MARK: - Streaming decoder

/// Incrementally decodes msgpack values from a byte stream.
///
/// Feed data with `append(_:)`, then extract decoded values with `nextValue()`.
final class StreamDecoder {
    private var buffer = Data()

    /// Append received bytes.
    func append(_ data: Data) {
        buffer.append(data)
    }

    /// Try to decode the next value from the buffer.
    /// Returns nil if there is not enough data yet.
    func nextValue() -> MessagePackValue? {
        guard !buffer.isEmpty else { return nil }
        do {
            let (value, remainder) = try unpack(buffer)
            buffer = Data(remainder)
            return value
        } catch {
            // Not enough data yet, or malformed -- wait for more.
            return nil
        }
    }
}
